addons/sms/models/sms_tracker.py

```diff
@@ -0,0 +1,87 @@
+    SMS through the SMS UUID field.
+    (Note: they are not tied to the SMS records directly as those are deleted when sent).
+    """
+    _name = 'sms.tracker'
+    _description = "Link SMS to mailing/sms tracking models"
+
+    SMS_STATE_TO_NOTIFICATION_STATUS = {
+        'canceled': 'canceled',
+        'process': 'process',
+        'error': 'exception',
+        'outgoing': 'ready',
+        'sent': 'pending',
+        'success': 'pending',
+        'delivered': 'sent',
+    }
+
+    sms_uuid = fields.Char('SMS uuid', required=True, index="btree_not_null")
+    mail_notification_id = fields.Many2one('mail.notification', ondelete='cascade')
+
+    _sql_constraints = [
+        ('sms_uuid_unique', 'unique(sms_uuid)', 'A record for this UUID already exists'),
+    ]
+
+    def _action_update_from_provider_error(self, provider_error):
+        """
+        :param str provider_error: value returned by third party (IAP) or any string.
+            If provided, notification values will be derived from it.
+            (see ``_get_tracker_values_from_provider_error``)
+        """
+        failure_reason = False
+        failure_type = f'sms_{provider_error}'
+        error_status = None
+        if failure_type not in self.env['sms.sms'].DELIVERY_ERRORS:
+            failure_type = 'unknown'
+            failure_reason = provider_error
+        elif failure_type in self.env['sms.sms'].BOUNCE_DELIVERY_ERRORS:
+            error_status = "bounce"
+
+        self._update_sms_notifications(error_status or 'exception', failure_type=failure_type, failure_reason=failure_reason)
+        return error_status, failure_type, failure_reason
+
+    def _action_update_from_sms_state(self, sms_state, failure_type=False):
+        notification_status = self.SMS_STATE_TO_NOTIFICATION_STATUS[sms_state]
+        if not notification_status:
+            raise ValueError(f'Unknown SMS state {sms_state}')
+        self._update_sms_notifications(notification_status, failure_type=failure_type)
+
+    def _update_sms_notifications(self, notification_status, failure_type=False, failure_reason=False):
+        if not self.mail_notification_id:  # avoid the search below
+            return
+
+        # canceled is a state which means that the SMS sending order should not be sent to the SMS service.
+        # 'process', 'pending' are sent to IAP which is not revertible (and 'sent' = 'delivered').
+        notifications_statuses_to_ignore = {
+            'canceled': ['canceled', 'process', 'pending', 'sent'],
+            'ready': ['ready', 'process', 'pending', 'sent'],
+            'process': ['process', 'pending', 'sent'],
+            'pending': ['pending', 'sent'],
+            'bounce': ['bounce', 'sent'],
+            'sent': ['sent'],
+            'exception': ['exception'],
+        }[notification_status]
+
+        notifications_sudo = self.env['mail.notification'].search([
```

Comment:
Use a filtered on self.mail_notification_id. Anyway I bet we are going to update most traces / notifications in a standard flow (they will follow responses from IAP) so not sure we will filter out much.